providers/defltprov.c

```diff
@@ -412,6 +412,11 @@ static const OSSL_ALGORITHM deflt_asym_cipher[] = {
 
 static const OSSL_ALGORITHM deflt_asym_kem[] = {
     { PROV_NAMES_RSA, "provider=default", ossl_rsa_asym_kem_functions },
+#ifndef OPENSSL_NO_EC
+    { PROV_NAMES_X25519, "provider=default", ossl_ecx_asym_kem_functions },
+    { PROV_NAMES_X448, "provider=default", ossl_ecx_asym_kem_functions },
+    { PROV_NAMES_EC, "provider=default", ossl_ec_asym_kem_functions },
```

Comment:
> Another possible naming approach could be to have a single algorithm called "DHKEM" and extend the "groups" that can be passed to include X25519 and X448 and well as normal EC groups.

That would be IMO very problematic from the implementation standpoint because the key types for these groups are different.

Reply:
> That would be IMO very problematic from the implementation standpoint because the key types for these groups are different.

Isn't that just a matter of special casing the X25519/X448 "groups". Although I suppose X25519 isn't really a group at all (Curve25519 is)...so maybe that doesn't work

Reply:
Thinking about it more - the current way how libcrypto implements the evp_kem_init() does not allow for specifying a different type of KEM on a given key type - what evp_keymgmt_util_query_operation_name() returns is used. Unless we really plan to extend libcrypto API to allow that, there is no point in changing the name of the KEM algorithm to confuse things, or we should have already done it with RSA.

Reply:
Hmmm. That seems like a fairly nasty limitation of our current API. We really should have some mechanism of determining which KEM algorithm we want to apply. Not least because if a provider decided to implement some other KEM for (say) EC, then suddenly HPKE is not going to work anymore. So, effectively we are locking in "DHKEM" as *the* KEM for use with EC/X25519/X448 forever.

But, ok, that's probably not a problem to be solved in *this* PR.

Reply:
> Isn't that just a matter of special casing the X25519/X448 "groups".

The point is that the keymgmt key object is passed to the operation in the provider. There is no way for the encapsulate_init to recognize what type is the key object unless the algorithm fetched is different.

Reply:
EVP_PKEY_CTX_set_kem_op() is the way we currently set the operation type.. If the init needs to be really different it can be defered to after the op is set. (This is a settable param so it can also be passed to the init).

I think before we go fixing 'groups' here we may want to consider if having a single EC keymanagement object to represent all curve types is a good idea or not.. I also think this is out of scope within this PR and needs a bigger discussion within OTC.
I initially had shared code to handle both EC and ECX and it was horrible and unreadable. Separating them out simplified the implementations (even if it is similiar looking code for each - it still has to handle the keys differently).

Reply:
> EVP_PKEY_CTX_set_kem_op() is the way we currently set the operation type.. If the init needs to be really different it can be defered to after the op is set. (This is a settable param so it can also be passed to the init).

Heh, interesting, I did not know about this function. So we basically have the algorithm choice as a parameter that can be freely ignored by the provider. IMO not too nice. Unfortunately we have to live with that.

IMO merging EC and ECX key types is not a good idea, and definitely not something that should be considered earlier than in 4.0.